<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Description" data-toc-modified-id="Description-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Description</a></span><ul class="toc-item"><li><span><a href="#The-goals-of-the-study" data-toc-modified-id="The-goals-of-the-study-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>The goals of the study</a></span></li><li><span><a href="#Access-to-the-data-base" data-toc-modified-id="Access-to-the-data-base-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Access to the data base</a></span></li><li><span><a href="#Study-the-tables-(print-the-first-rows)" data-toc-modified-id="Study-the-tables-(print-the-first-rows)-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Study the tables (print the first rows)</a></span></li></ul></li><li><span><a href="#Tasks" data-toc-modified-id="Tasks-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Tasks</a></span><ul class="toc-item"><li><span><a href="#Find-the-number-of-books-released-after-January-1,-2000." data-toc-modified-id="Find-the-number-of-books-released-after-January-1,-2000.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Find the number of books released after January 1, 2000.</a></span></li><li><span><a href="#Find-the-number-of-user-reviews-and-the-average-rating-for-each-book." data-toc-modified-id="Find-the-number-of-user-reviews-and-the-average-rating-for-each-book.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Find the number of user reviews and the average rating for each book.</a></span></li><li><span><a href="#Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages." data-toc-modified-id="Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Identify the publisher that has released the greatest number of books with more than 50 pages.</a></span></li><li><span><a href="#Identify-the-author-with-the-highest-average-book-rating-(look-only-at-books-with-at-least-50-ratings)." data-toc-modified-id="Identify-the-author-with-the-highest-average-book-rating-(look-only-at-books-with-at-least-50-ratings).-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Identify the author with the highest average book rating (look only at books with at least 50 ratings).</a></span></li><li><span><a href="#Find-the-average-number-of-text-reviews-among-users-who-rated-more-than-50-books." data-toc-modified-id="Find-the-average-number-of-text-reviews-among-users-who-rated-more-than-50-books.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Find the average number of text reviews among users who rated more than 50 books.</a></span></li></ul></li><li><span><a href="#Conclusion." data-toc-modified-id="Conclusion.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Conclusion.</a></span></li></ul></div>

# Developing a value proposition for a new product SQL

## Description

![](Untitled_-_2020-07-02T142019.920_1593688954.png)

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.
Description of the data

<b>books:</b>

Contains data on books:

`book_id`

`author_id`

`title`

`num_pages` — number of pages

`publication_date`

`publisher_id`

<b>authors:</b>

Contains data on authors:

`author_id`

`author`

<b>publishers:</b>

Contains data on publishers:

`publisher_id`

`publisher`

<b>ratings:</b>

Contains data on user ratings:

`rating_id`

`book_id`

`username` — the name of the user who rated the book

`rating`
    
<b>reviews:</b>

Contains data on customer reviews:

`review_id`

`book_id`

`username` — the name of the user who reviewed the book

`text` — the text of the review

### The goals of the study

Generate information for making a value proposition for a new product.

### Access to the data base

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Study the tables (print the first rows)

In [3]:
books = pd.io.sql.read_sql(''' SELECT * FROM books LIMIT 5 ''', con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
authors = pd.io.sql.read_sql(''' SELECT * FROM authors LIMIT 5 ''', con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
publishers = pd.io.sql.read_sql(''' SELECT * FROM publishers LIMIT 5 ''', con = engine)
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
ratings = pd.io.sql.read_sql(''' SELECT * FROM ratings LIMIT 5 ''', con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
reviews = pd.io.sql.read_sql(''' SELECT * FROM reviews LIMIT 5 ''', con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Tasks

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

### Find the number of books released after January 1, 2000.

In [8]:
query_b = f'''SELECT COUNT(book_id) FROM books AS books_number'''
books_total = pd.read_sql(query_b, engine)
books_total

count
0   1000

In [9]:
books_rel = f''' SELECT COUNT(book_id) FROM books AS books_number
                 WHERE publication_date > '2000-01-01' '''
books__since_2000 = pd.read_sql(books_rel, engine)
books__since_2000

count
0    819

Conclusion: Total number of books in the dataset is 1000, 819 of them released after January 1, 2000.

### Find the number of user reviews and the average rating for each book.

In [10]:
query_rew = f'''SELECT COUNT(review_id) FROM reviews AS review_number'''
reviews_total = pd.read_sql(query_rew, engine)
reviews_total

count
0   2793

In [11]:
query_avr2 = f''' SELECT COUNT(DISTINCT review_id) AS reviews_number, AVG(rating) AS avg_rating, books.book_id, title
            FROM(( books
            LEFT OUTER JOIN ratings ON ratings.book_id = books.book_id)
            LEFT OUTER JOIN reviews ON reviews.book_id = ratings.book_id)
            GROUP BY books.book_id
            ORDER BY avg_rating ASC '''

book_avr2 = pd.read_sql(query_avr2, engine)
book_avr2

reviews_number  avg_rating  book_id  \
0                 2        1.50      303   
1                 2        2.00      371   
2                 2        2.00      316   
3                 3        2.00      202   
4                 3        2.25      915   
..              ...         ...      ...   
995               2        5.00      421   
996               2        5.00      418   
997               2        5.00      169   
998               2        5.00      901   
999               2        5.00      902   

                                                 title  
0                                 Harvesting the Heart  
1                                                Junky  
2                    His Excellency: George Washington  
3                                        Drowning Ruth  
4    The World Is Flat: A Brief History of the Twen...  
..                                                 ...  
995                                        Marvel 1602  
996                                              March  
997  Crucial Conversations: Tools for Talking When ...  
998  The Walking Dead  Book One (The Walking Dead #...  
999  The War of Art: Break Through the Blocks & Win...  

[1000 rows x 4 columns]

Conclusion: Data contains 1000 books with at least one review and average rating from 1,5 to 5.

### Identify the publisher that has released the greatest number of books with more than 50 pages.

In [12]:
query_pub = f''' SELECT publisher, COUNT(books.book_id) FROM publishers
                 INNER JOIN books ON books.publisher_id = publishers.publisher_id
                 WHERE num_pages > 50
                 GROUP BY publishers.publisher
                 ORDER BY COUNT(book_id) DESC '''
pub_50 = pd.read_sql(query_pub, engine)
pub_50

publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]

Conclusion: Publisher "Penguin Books" released the greatest number of books with more than 50 pages (42 books).

### Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [18]:
pd.read_sql(''' SELECT author, r.avg
                FROM (SELECT b.book_id AS id, b.author_id as author_id,
                COUNT(DISTINCT(rat.rating_id)),
                AVG(rat.rating) as avg
                FROM books AS b
                INNER JOIN ratings as rat ON b.book_id = rat.book_id
                GROUP BY id HAVING COUNT(DISTINCT(rat.rating_id)) > 50) as r
                INNER JOIN authors as a ON a.author_id = r.author_id
                GROUP BY author, r.avg
                ORDER BY r.avg DESC
                '''
                , engine)


author       avg
0                          J.K. Rowling/Mary GrandPré  4.414634
1                                      J.R.R. Tolkien  4.391892
2                          J.K. Rowling/Mary GrandPré  4.287500
3                   Markus Zusak/Cao Xuân Việt Khương  4.264151
4                          J.K. Rowling/Mary GrandPré  4.246575
5                                   Louisa May Alcott  4.192308
6                          J.K. Rowling/Mary GrandPré  4.186667
7                                      J.R.R. Tolkien  4.125000
8                                        Rick Riordan  4.080645
9                                     William Golding  3.901408
10                                          Dan Brown  3.830508
11                                      J.D. Salinger  3.825581
12           Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
13  William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
14                                         Lois Lowry  3.750000
15            George Orwell/Boris Grabnar/Peter Škerl  3.729730
16                                          Dan Brown  3.678571
17                                    Stephenie Meyer  3.662500
18                                     John Steinbeck  3.622951

In [24]:
pd.read_sql(''' 

WITH book_ratings AS
  (SELECT author,
          r.avg
   FROM
     (SELECT b.book_id AS id,
             b.author_id AS author_id,
             COUNT(DISTINCT(rat.rating_id)),
             AVG(rat.rating) AS AVG
      FROM books AS b
      INNER JOIN ratings AS rat ON b.book_id = rat.book_id
      GROUP BY id
      HAVING COUNT(DISTINCT(rat.rating_id)) > 50) AS r
   INNER JOIN authors AS a ON a.author_id = r.author_id
   GROUP BY author,
            r.avg)
SELECT author,
       COUNT(author) AS book_num,
       AVG(AVG)
FROM book_ratings
GROUP BY author
ORDER BY AVG DESC

''', engine)

author  book_num       avg
0                          J.K. Rowling/Mary GrandPré         4  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương         1  4.264151
2                                      J.R.R. Tolkien         2  4.258446
3                                   Louisa May Alcott         1  4.192308
4                                        Rick Riordan         1  4.080645
5                                     William Golding         1  3.901408
6                                       J.D. Salinger         1  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce         1  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...         1  3.787879
9                                           Dan Brown         2  3.754540
10                                         Lois Lowry         1  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl         1  3.729730
12                                    Stephenie Meyer         1  3.662500
13                                     John Steinbeck         1  3.622951

Conclusion: J.K. Rowling/Mary GrandPré the author with the highest average book rating (4.43), which has at least 50 ratings per book.

### Find the average number of text reviews among users who rated more than 50 books.

In [14]:
query_rev = f''' SELECT AVG(review_number) FROM (SELECT 
                 COUNT(DISTINCT reviews.review_id) as review_number
                 FROM ratings INNER JOIN reviews ON reviews.username = ratings.username
                 GROUP BY reviews.username 
                 HAVING COUNT(DISTINCT ratings.rating_id) > 50) AS q '''
avr_rev_num = pd.read_sql(query_rev, engine)
avr_rev_num

avg
0  24.333333

Conclusion: the average number of text reviews among users who rated more than 50 books is 24.

## Conclusion.

Total number of books in the dataset is 1000, 819 of them released after January 1, 2000.

Books has at least one review and average rating from 1,5 to 5.

Publisher "Penguin Books" released the greatest number of books with more than 50 pages (42 books).

J.K. Rowling/Mary GrandPré the author with the highest average book rating (4.3), which has at least 50 ratings per book.

The average number of text reviews among users who rated more than 50 books is 24.